pyhms/demes/single_pop_eas/de.py

```diff
@@ -0,0 +1,210 @@
+            else np.argsort(-1 * population.fitnesses)
+        )
+        p_best = []
+        for p_i in p:
+            best_index = sorted_fitness_indexes[: max(2, int(round(p_i * population.size)))]
+            p_best.append(np.random.choice(best_index))
+        p_best_np = np.array(p_best)
+        randoms = get_randoms(population)
+        randoms_with_archive = get_randoms(population.merge(archive)) if archive is not None else randoms
+        randoms_with_archive = randoms_with_archive[: len(randoms)]
+        mutated_genomes = (
+            population.genomes
+            + f * (population.genomes[p_best_np] - population.genomes)
+            + f * (randoms[:, 0] - randoms_with_archive[:, 1])
+        )
+        new_genomes = apply_bounds(mutated_genomes, population.problem.bounds, "reflect")
+        new_fitness = np.where(
+            np.all(new_genomes == population.genomes, axis=1),
+            population.fitnesses,
+            np.nan,
+        )
+        return Population(new_genomes, new_fitness, population.problem)
+
+
+class Crossover:
+    def __call__(
+        self,
+        population: Population,
+        mutated_population: Population,
+        probability: np.ndarray | float,
+    ) -> Population:
+        chosen = np.random.rand(*population.genomes.shape)
+        j_rand = np.random.randint(0, population.genomes.shape[1])
+        chosen[j_rand :: population.genomes.shape[1]] = 0  # noqa: E203
+        new_genomes = np.where(chosen <= probability, mutated_population.genomes, population.genomes)
+        new_fitness = np.where(
+            np.all(new_genomes == population.genomes, axis=1),
+            population.fitnesses,
+            np.nan,
+        )
+        new_population = Population(new_genomes, new_fitness, population.problem)
+        return new_population
+
+
+class DE:
+    def __init__(self, use_dither: bool, crossover_probability: float, f: float | None = None) -> None:
+        self._crossover_probability = crossover_probability
+        self._mutation = BinaryMutationWithDither() if use_dither else BinaryMutation(f=f)
+        self._crossover = Crossover()
+
+    def run(self, parents: list[Individual]) -> list[Individual]:
+        parent_population = Population.from_individuals(parents)
+        trial_population = parent_population.copy()
+        trial_population = self._mutation(trial_population)
+        trial_population = self._crossover(parent_population, trial_population, self._crossover_probability)
+        trial_population.evaluate()
+        new_population_indices = (
+            (trial_population.fitnesses >= parent_population.fitnesses)
+            if parent_population.problem.maximize
+            else (trial_population.fitnesses <= parent_population.fitnesses)
+        )
+        return (
+            trial_population[new_population_indices].merge(parent_population[~new_population_indices]).to_individuals()
+        )
+
+
+class SHADE:
```

Comment:
SHADE implementation is the biggest and most advanced part of this pull request. Honestly, I would like to talk it through with you for better inspection of this part.